lib/Service/FontService.php

```diff
@@ -0,0 +1,250 @@
+<?php
+/**
+ * @copyright Copyright (c) 2022 Julien Veyssier <eneiluj@posteo.net>
+ *
+ * @author Julien Veyssier <eneiluj@posteo.net>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Richdocuments\Service;
+
+
+use OCA\Richdocuments\AppInfo\Application;
+use OCP\Files\IAppData;
+use OCP\Files\NotFoundException;
+use OCP\Files\SimpleFS\ISimpleFile;
+use OCP\Files\SimpleFS\ISimpleFolder;
+use OCP\ICacheFactory;
+use OCP\IConfig;
+use OCP\IURLGenerator;
+
+class FontService {
+	private const INVALIDATE_FONT_LIST_CACHE_AFTER_SECONDS = 3600;
+
+	/**
+	 * @var IAppData
+	 */
+	private $appData;
+	/**
+	 * @var \OCP\ICache
+	 */
+	private $cache;
+	/**
+	 * @var IURLGenerator
+	 */
+	private $url;
+	/**
+	 * @var IConfig
+	 */
+	private $config;
+
+	public function __construct(IAppData $appData,
+								ICacheFactory $cacheFactory,
+								IURLGenerator $url,
+								IConfig $config) {
+		$this->appData = $appData;
+		$this->cache = $cacheFactory->createDistributed(Application::APPNAME);
+		$this->url = $url;
+		$this->config = $config;
+	}
+
+	/**
+	 * @return ISimpleFolder
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getFontAppDataDir(): ISimpleFolder {
+		try {
+			return $this->appData->getFolder('fonts');
+		} catch (NotFoundException $e) {
+			return $this->appData->newFolder('fonts');
+		}
+	}
+
+	/**
+	 * @return ISimpleFolder
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getFontOverviewAppDataDir(): ISimpleFolder {
+		try {
+			return $this->appData->getFolder('font-overviews');
+		} catch (NotFoundException $e) {
+			return $this->appData->newFolder('font-overviews');
+		}
+	}
+
+	/**
+	 * Get the list of available font files
+	 *
+	 * @return array
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function getFontFiles(): array {
+		$cacheKey = 'fontFiles';
+		$cachedFiles = $this->cache->get($cacheKey);
+		if ($cachedFiles === null) {
+			$fontDir = $this->getFontAppDataDir();
+			$cachedFiles = $fontDir->getDirectoryListing();
+			$this->cache->set($cacheKey, $cachedFiles, self::INVALIDATE_FONT_LIST_CACHE_AFTER_SECONDS);
```

Comment:
With APCu used as a cache the serialisation does of ISimpleFile objects doesn't work as expected here.

```
Exception
0
Serialization of 'Closure' is not allowed
/var/www/html/lib/private/Memcache/APCu.php
49

Trace

#0 /var/www/html/lib/private/Memcache/APCu.php(49): apcu_store('47f7f58ed821567...', Array, 3600)
#1 /var/www/html/apps-extra/richdocuments/lib/Service/FontService.php(102): OC\Memcache\APCu->set('fontFiles', Array, 3600)
#2 /var/www/html/apps-extra/richdocuments/lib/Controller/SettingsController.php(338): OCA\Richdocuments\Service\FontService->getFontFiles()
#3 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(225): OCA\Richdocuments\Controller\SettingsController->getJsonFontList()
#4 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(133): OC\AppFramework\Http\Dispatcher->executeController(Object(OCA\Richdocuments\Controller\SettingsController), 'getJsonFontList')
#5 /var/www/html/lib/private/AppFramework/App.php(172): OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\Richdocuments\Controller\SettingsController), 'getJsonFontList')
#6 /var/www/html/lib/private/Route/Router.php(298): OC\AppFramework\App::main('OCA\\Richdocumen...', 'getJsonFontList', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)
#7 /var/www/html/lib/base.php(1023): OC\Route\Router->match('/apps/richdocum...')
#8 /var/www/html/index.php(36): OC::handleRequest()
#9 {main}
```

Reply:
This is useless to cache the file list anyway. This list is computed when
* getting the file name list (which is already cached)
* reaching the json font endpoint which is not supposed to happen too often

No more cache in `FontService::getFontFiles()`